rasa/nlu/model.py

```diff
@@ -208,46 +210,134 @@ def train(self, data: TrainingData, **kwargs: Any) -> "Interpreter":
 
+        data_chunk_dir = tempfile.mkdtemp()
+        dir_name, model_name = self._create_model_dir(path, fixed_model_name)
+
+        # perform tokenization & prepare other components for training in chunks
+        for i, component in enumerate(self.pipeline):
+            if isinstance(component, Tokenizer):
+                component.train(training_data, self.config, **context)
+                metadata["pipeline"].append(
+                    self._persist_component(component, dir_name, i)
+                )
+            else:
+                component.prepare_partial_training(training_data)
+
+        training_data_chunks = training_data.divide_into_chunks()
+        number_of_chunks = len(training_data_chunks)
+
+        # perform featurization
+        for i, data_chunk in enumerate(training_data_chunks):
+            for component in self.pipeline:
+                if isinstance(component, Featurizer):
+                    component.train_chunk(data_chunk, self.config, **context)
+            data_chunk.persist_chunk(data_chunk_dir, f"{i}_chunk.porto")
+
+        # persist featurizers
+        for i, component in enumerate(self.pipeline):
+            if isinstance(component, Featurizer):
+                metadata["pipeline"].append(
+                    self._persist_component(component, dir_name, i)
+                )
+
+        # TODO training of classifiers probably needs to be adapted
+        for i, component in enumerate(self.pipeline):
+            if isinstance(component, (IntentClassifier, EntityExtractor)):
+                for j in range(number_of_chunks):
+                    file_path = os.path.join(data_chunk_dir, f"{j}_chunk.porto")
+                    data_chunk = TrainingDataChunk.load_chunk(file_path)
+                    component.train_chunk(data_chunk, self.config, **context)
+                metadata["pipeline"].append(
+                    self._persist_component(component, dir_name, i)
+                )
+
+        Metadata(metadata, dir_name).persist(dir_name)
+
+        if persistor is not None:
+            persistor.persist(dir_name, model_name)
+        logger.info(
+            "Successfully saved model into '{}'".format(os.path.abspath(dir_name))
+        )
+
+        return Interpreter(self.pipeline, context), dir_name
+
+    def _create_model_dir(self, path: Text, fixed_model_name: Text = None):
+        timestamp = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+
         if fixed_model_name:
             model_name = fixed_model_name
         else:
             model_name = NLU_MODEL_NAME_PREFIX + timestamp
 
         path = os.path.abspath(path)
         dir_name = os.path.join(path, model_name)
-
         rasa.shared.utils.io.create_directory(dir_name)
 
+        return dir_name, model_name
+
+    @staticmethod
+    def _file_name(index: int, name: Text) -> Text:
+        return f"component_{index}_{name}"
+
+    def persist(
+        self,
+        path: Text,
```

Comment:
As this command is implemented from scratch it would be awesome if we could switch to use `pathlib` instead of `Text` paths everywhere unless it's an enormous pain to hook it up with the legacy code.